### ISSUE TITLE
Support custom fixtures

### ISSUE THREAD

**Issue by cassus** [vitalets/playwright-bdd#11]
Thank you for starting this project, it's great to see progress on the playwright-bdd front!

I'm wondering how would it be possible to use custom fixtures in the steps. This would enable component testing (`mount`), or other custom fixtures I'm using in my project (especially worker fixtures which are a great performance optimization)

**Comment by vitalets** [comment]
This is a very good topic :)

Currently for **test-scoped fixtures** it's possible to use CustomWorld with lazy getters. For example:
```ts
type User = {
  id: string;
  name: string;
  password: string;
}

export class CustomWorld extends World {
  private user: User;

  async getUser() {
    if (!this.user) {
         this.user = await this.request.post('/api/createUser');
    }

    return this.user;
  }

  async destroy() {
    if (this.user) {
        await this.request.post(`/api/deleteUser/${this.user.id}`);
    }
  }
}

setWorldConstructor(CustomWorld);
```
Then in step definitions:
```ts
When('sign in', async function (this: World) {
  const user = await this.getUser();
  await this.page.getByLabel('User Name').fill(user.name);
  await this.page.getByLabel('Password').fill(user.password);
  await this.page.getByText('Sign in').click();
});

Then('username is in header', async function (this: World) {
  const user = await this.getUser();
  await expect(this.page.getByRole('header')).toContainText(user.name);
});
```
The downside is that syntax is not so concise as for fixtures. Compare to:
```ts
test.extend({
  user: async ({ request }, use) => {
    const user = await request.post('/api/createUser');
    await use(user);
    await request.post(`/api/deleteUser/${this.user.id}`);
  },
});
```

**Comment by vitalets** [comment]
For **worker-scoped fixtures** there is no way for now. 
But I suggest to spin discussion on that as technically it's possible, we just need to find a proper way.

Imagine the following worker-scoped fixture:
```ts
test.extend({
  user: [async ({ request }, use) => {
    const user = await request.post('/api/createUser');
    await use(user);
    await request.post(`/api/deleteUser/${this.user.id}`);
  }, { scope: 'worker' }],
});
```

Open questions:
1. How from Gherkin syntax can we detect that some step uses the `user` fixture?
   ```gherkin
   Feature: scenario-simple
      # How to indicate usage of `user` fixture?
      Scenario: Scenario one
        Given State 1
        When Action 1
   ```
2. Maybe indicate usage of fixture from step definition? But function args are reserved for Cucumber parameters. We can attach something to World:
   ```ts
   When('sign in', async function (this: World) {
      const user = this.use('user'); // <-- use of fixture?
      await this.page.getByLabel('User Name').fill(user.name);
      await this.page.getByLabel('Password').fill(user.password);
      await this.page.getByText('Sign in').click();
    });
   ```
3. How to pass `user` fixture to step definition?
   ```ts
   test("Scenario one", async ({ Given, When, user }) => {
     await Given("State 1", { user }); // <-- pass user here?
     await When("Action 1");
   });
   ```

Feel free to share your ideas on that =)

**Comment by mortelee** [comment]
On question 3: If we would create multiple fixtures how would your spec generation know on which test to add a specific fixture?

**Comment by vitalets** [comment]
> On question 3: If we would create multiple fixtures how would your spec generation know on which test to add a specific fixture?

The only way I have in mind - load step definitions during generation and look for `this.use('xxx')` substring in step function body. Not fully like such substring searing but no more ideas for now.

Ideally, step definitions could have the same fixture destructuring argument as Playwright's test function:
```ts
When('sign in', async function ({ world, user, page }) {
   // ...
});
```
But this way we are totally breaking Cucumber compatibility.

**Comment by vitalets** [comment]
> I'm wondering if there is a functional (vs. OO) alternative to Cucumber that we could use as a starting point. I don't know of any, just wondering.

I also don't aware of such alternatives, Cucumber is de-facto standard. I think it's object oriented mainly for having World context. But if we use Playwright fixtures it seems to me that World becomes unnecessary. 

What we can do - is to provide own `Given / When / Then` functions that will internally call Cucumber's ones but allows to use fixtures. 

For example:
```ts
import { Given } from 'playwright-bdd';

Given('I open url {string}', async ({ page }, url: string) => {
   await page.goto(url);
});
```

What do you think?

**Comment by mortelee** [comment]
If we want to add our own fixtures we need to extend test from "playwright-bdd". How to make sure that our own test function becomes available in generated spec files?

**Comment by cassus** [comment]
> `Given('I open url {string}', async ({ page }, url: string) => {` 

I like it a lot! It's a shift form a _cucumber playwright translator tool_ to a _playwright BDD tool_ that uses cucumber under the hood. 

Using fixtures instead of the World object makes me feel much more at home in a playwright codebase. As an escape hatch we can always create a _world_ fixture that contains the world object.

> If we want to add our own fixtures we need to extend test from "playwright-bdd". How to make sure that our own test function becomes available in generated spec files?

I think the same playwright `test.extend` pattern would work well here.
Eg.
### fixtures.js
```ts
import { test as baseTest } from 'playwright-bdd'
const test = test.extend(...)
export const {Given, When, Then} = test
```

### test.js
```ts
import { Given } from './fixtures.js'
```

**Comment by mortelee** [comment]
You want to use playwright-bdd as a package. If I want to add new fixtures I don't want to update fixtures.js inside the package.
I want to be able to use my own fixture file (that is part of my test project) when generating spec files.

**Comment by NikkTod** [comment]
+1 on this one, I am trying to migrate to BDD the following test (using POM with multiple custom fixtures for the pages) 

![image](https://github.com/vitalets/playwright-bdd/assets/94455079/ebb94484-3e57-4747-8196-30b38de78a2f)

**Comment by vitalets** [comment]
> +1 on this one, I am trying to migrate to BDD the following test (using POM with multiple custom fixtures for the pages)

Working on it!

Btw, do you have cases when there are several different `test.extend()` calls? I mean there are different set of custom fixtures that you are applying for different tests? Or it's a single place where you call `test.extend` with all custom fixtures?

**Comment by cassus** [comment]
https://github.com/Niitch/gherkin-wrapper/issues/1 Has a great solution for having worker and test fixtures available in the steps

**Comment by vitalets** [comment]
> Yes I do, have layers of `test.extend()`:
> 
> * Fixtures common across e2e and component tests
> * Fixtures common across e2e tests
> * Fixtures common across a feature area
> * Fixtures specific for a test file

Got it, thanks.

> https://github.com/Niitch/gherkin-wrapper/issues/1 Has a great solution for having worker and test fixtures available in the steps

Yes, I've looked on it. We are achieving the same goal with a bit different approach. `gherkin-wrapper` converts feature files on-the-fly, `playwright-bdd` converts them on pre-run providing generated test files (mainly for test debugging purpose). I think both approaches are viable and eventually we will cover all pros and cons for playwright team to integrate bdd natively.

**Comment by vitalets** [comment]
Prepared the next release with **custom fixtures** support.
I would appreciate if you try it. 
@cassus @mortelee @NikkTod 

Install `next` version:
```
npm i -D playwright-bdd@next
```

Check steps with default fixtures (`page`):
```ts
import { createBDD } from 'playwright-bdd';

const { Given, When, Then } = createBDD();

Given('I open url {string}', async ({ page }, url: string) {
  await page.goto(url);
});
```

Check steps with custom fixtures:
```ts
import { createBDD } from 'playwright-bdd';
import { Page } from '@playwright/test';

// test-scoped fixture
class MyPage {
  constructor(public page: Page) {}

  async openLink(name: string) {
    await this.page.getByRole('link', { name }).click();
  }
}

// pass custom fixtures to createBDD() similar to test.extend()
const { Given, When, Then } = createBDD<{ myPage: MyPage }>({
  myPage: async ({ page }, use) => {
    await use(new MyPage(page));
  }
});

// use fixtures in steps
Given('I open url {string}', async ({ page }, url: string) => {
  await page.goto(url);
});

When('I click link {string}', async ({ myPage }, name: string) => {
  await myPage.openLink(name);
});
```

Detailed documentation is in [custom-fixtures](https://github.com/vitalets/playwright-bdd/tree/custom-fixtures#writing-steps) branch.
Let me know if there are any problems.

**Comment by NikkTod** [comment]
@vitalets I see changes are in custum-features branch, should we wait until you merge to main, or we just run the 'npm i -D playwright-bdd@next'

**Comment by vitalets** [comment]
> @vitalets I see changes are in custum-features branch, should we wait until you merge to main, or we just run the 'npm i -D playwright-bdd@next'

Yes, you can run `npm i -D playwright-bdd@next` and check it before release.

**Comment by vitalets** [comment]
> What is the benefit of implementing a separate extension logic like test.extend instead of just accepting a "test" instance?

Well, this is because of async loading of step definitions. In generated test files we need to have `test` fn with all needed fixtures on the first tick, but step definitions are not loaded at that moment. That's why I need to store user's fixtures and provide them to `test.extend` "later". I'd like to keep using `test.extend` directly, but didn't find such solution.

> Is there a way to use the "mount" hook for component tests in this way?

I didn't try component testing yet, but I will try.

**Comment by NikkTod** [comment]
Hey @vitalets, good work on the custom features. I have tested it with multiple test-scoped features together with page object model and it work great. 

Few things to mention:
1. Is it possible to add a bit of formating on the generated file when you extend test with the custom features? For example if we have multiple features it will be good if each one is moved on new line.
2. If we have an error (like in the screenshot) is it possible that error to be on the line/step where that error occured?
![image](https://github.com/vitalets/playwright-bdd/assets/94455079/249abe94-d2cf-4de0-896f-30c4826e3737)

3. Auto-generation of the playwright test file on certain actions:
 - Well I dont know if this is even possible at all, but it would be very nice if we can auto re-generate the playwright test files on changes within the feature file. Or somehow maybe if upon saving event in VS Code. So that when we want to start tests locally( from VS code extention play icon) we do not need to regenerate the test files after every change in the feature file.
 - Another example is if I open UI mode and click on the watch icon for a particular test, it will watch only for changes in the generated playwright test files , so if you update a step file or a feaure file this will not trigger the test run.
 
I am now waiting to get this feature of wrapping each cucumber step in test.step. :)

**Comment by vitalets** [comment]
Hi @NikkTod thanks for the feedback.

> 1. Is it possible to add a bit of formating on the generated file when you extend test with the custom features? For example if we have multiple features it will be good if each one is moved on new line.

It is done intentionally as this line of generated code is kind of technical stuff. To keep focus on tests itself. Anyway after feedback of @casus I have an idea how to avoid this line at all. I'm thinking about adding a parameter `importTestFrom` to point to test exporting file. If provided this file will be used to import `test` instead of `import { test } from 'playwright-bdd`. I allows to keep using `test.extend()` function as we do in not-bdd playwright tests.


> 2. If we have an error (like in the screenshot) is it possible that error to be on the line/step where that error occured?

I will check. Is it works correctly on current non custom version of playwright-bdd? If also not could you create a separate issue for that?

> 3. Auto-generation of the playwright test file on certain actions

Could you check [watch mode](https://github.com/vitalets/playwright-bdd#watch-mode)?


> I am now waiting to get this feature of wrapping each cucumber step in test.step. :)

This will be done in https://github.com/vitalets/playwright-bdd/issues/22. I've checked locally, reporting becomes I very pretty with it.

**Comment by vitalets** [comment]
Finally implemented custom fixtures in v3.0.1. Documented new syntax in README.
@cassus @mortelee @NikkTod I would appreciate your feedback :)
